### DISCUSSION TITLE
vsgCs 0.4

### DISCUSSION THREAD

**Discussion by timoore** [vsg-dev/VulkanSceneGraph#849] [General]
I just made a 0.4 release of [vsgCs](https://github.com/timoore/vsgCs), a library for displaying 3D Tiles and other Cesium ion content with VSG. 0.4 has these new features:

* The build process is more self-contained and will now download Cesiuam native libraries and VSG.
* vsgCs can stream Google Maps Photorealistic 3D Tiles.
* Some support for tile styling has been added. If present, the cesium#color feature attribute is used to color tiles.
* The management of web connections and data download has been greatly improved.
* The mouse manipulator from the [rocky](https://github.com/pelicanmapping/rocky) project has been integrated. This provides better interactive control, especially close to the Earth, than the default VSG trackball manipulator.
* Support for the [Tracy](https://github.com/wolfpld/tracy) profiling package has been added.
* The credits display has been rewritten, using images and attribution from the data sources.
* A "time of day" argument has been added to the worldviewer application, which is used in positioning the default light i.e., the Sun.
* Point clouds in 3D tiles can be rendered.

Enjoy!
Tim

**Comment by vsg-dev** [comment]
Congrats on the release.

**Comment by remoe** [comment]
@timoore , about Tracy: Vulkan GPU Tracing is not integrated, right?

**Comment by timoore** [comment]
> @timoore , about Tracy: Vulkan GPU Tracing is not integrated, right?

That is correct. I had thought that I would need to insert Tracy macros into VSG code to make that possible. I think I see a way to instrument vsgCs drawing, but it's not implemented yet.

**Reply by robertosfield** [reply]
I am not familiar with Tracy but if there is Vulkan timing support within it perhaps you could create a vsg::Command subclass that invokes the start and end timing?

There is query support supported by the VSG already:

https://github.com/vsg-dev/VulkanSceneGraph/blob/master/include/vsg/commands/BeginQuery.h
https://github.com/vsg-dev/VulkanSceneGraph/blob/master/include/vsg/commands/EndQuery.h

**Reply by remoe** [reply]
Thanks! Robert, I'm not familiar with VSG :) Still learning. 
But I've tried the following:

```c++
class TracyBeginCmd : public vsg::Inherit<vsg::BeginQuery, TracyBeginCmd>
{
public:
    TracyBeginCmd(vsg::Device* device) : _device(device) {}

    void record(vsg::CommandBuffer& commandBuffer) const override
    {
        auto queueFamily = _device->getPhysicalDevice()->getQueueFamily(VK_QUEUE_GRAPHICS_BIT);
        auto tracy_context = TracyVkContext(
            _device->getPhysicalDevice()->vk(),
            _device->vk(),
            _device->getQueue(queueFamily)->vk(),
            commandBuffer.vk());
    }
private:
    vsg::Device* _device;
};

main()
{
...
    auto commandGraph = vsg::createCommandGraphForView(window, camera, scene);
    commandGraph->addChild(TracyBeginCmd::create(window->getDevice()));
...
}
```

This results in the following error:
```
VUID-vkBeginCommandBuffer-commandBuffer-00049(ERROR / SPEC): msgNum: -2080204129 - Validation Error: [ VUID-vkBeginCommandBuffer-commandBuffer-00049 ] Object 0: handle = 0x2706c973310, type = VK_OBJECT_TYPE_COMMAND_BUFFER; | MessageID = 0x84029a9f | vkBeginCommandBuffer(): Cannot call Begin on VkCommandBuffer 0x2706c973310[] in the RECORDING state. Must first call vkEndCommandBuffer(). The Vulkan spec states: commandBuffer must not be in the recording or pending state (https://vulkan.lunarg.com/doc/view/1.3.243.0/windows/1.3-extensions/vkspec.html#VUID-vkBeginCommandBuffer-commandBuffer-00049)
    Objects: 1
       [0]  0x2706c973310, type: 6, name: NULL
```

Tracy docs: https://github.com/wolfpld/tracy/blob/cf0e6b610b02fbcbbf309b2de38a5e95604d0198/manual/tracy.tex#L1470
or the PDF: https://github.com/wolfpld/tracy/releases/latest/download/tracy.pdf

**Reply by robertosfield** [reply]
I'm away from my dev system so just a quick reply.  I presume the TracyVkContext call is trying to do stuff that is inappropriate for usage in a Command::record() implementation, perhaps is more appropriate for a Command::compile(Context&) implementation.

For the code you have provided you subclass from BeginQuery and override record(..), which may not be appropriate.  Subclassing from Command might be better, the overriding the compile(..) and record(..). However, it might be that Tracy will do all that.

It may be worth having a look at the vsgtimestates example as this illustrates the query interface:
https://github.com/vsg-dev/vsgExamples/blob/master/examples/commands/vsgtimestamps/vsgtimestamps.cpp

**Comment by timoore** [comment]
From my reading of the Tracy docs, TracyVkContext should be called once per
logical device and queue. The resulting context should be passed to the
TracyVkZone macro that marks the GPU zone.

The Tracy zone macros work via C++ RAII, i.e. a zone begins when an object
is created and ends at when it is destroyed (goes out of scope). There is a
C interface for CPU Tracy so that zones can begin and end in different
functions, but I don't see that this has been implemented for graphical
tracing. For tracing VSG code without changing VSG source, I would take the
approach of subclassing interesting VSG classes and overriding the record()
method to mark a Tracy zone and then call the original class.

Tim

**Reply by robertosfield** [reply]
Perhaps one could provide a couple of key subclasses like from vsg::StateCommand, vsg::Command and vsg::Node then implement them as decorator such that each subclass "has a" stateCommand, command, node then overrides the accept method and does then calls the accept on the decorated child.

This approach would allow one to decorate elements in the scene graph with the CPU and GPU tracing code where appropriate.  Perhaps one could even create a vsgTracy library to provide these subclasses.
